src/gwt/src/org/rstudio/core/client/FocusMap.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * FocusMap.java
+ *
+ * Copyright (C) 2009-16 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+package org.rstudio.core.client;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import org.rstudio.core.client.widget.CanFocus;
+
+import com.google.gwt.event.dom.client.KeyCodes;
+import com.google.gwt.event.dom.client.KeyDownEvent;
+import com.google.gwt.event.dom.client.KeyDownHandler;
+import com.google.gwt.user.client.ui.Focusable;
+import com.google.gwt.user.client.ui.Widget;
+
+public class FocusMap
+{
+   public FocusMap()
+   {
+      fwd_ = new HashMap<Widget, Widget>();
+      bwd_ = new HashMap<Widget, Widget>();
+      registration_ = new HashSet<Widget>();
+   }
+
+   public final void add(Widget source, Widget target)
+   {
+      fwd_.put(source, target);
```

Comment:
nit: Duplicates information a little, maybe use a flat LinkedList? (That has some other tradeoffs though so I'm fine with what you have.)